apps/playwright-browser-tunnel/src/PlaywrightBrowserTunnel.ts

```diff
@@ -0,0 +1,524 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import type { ChildProcess } from 'node:child_process';
+
+import type { BrowserServer, BrowserType, LaunchOptions } from 'playwright-core';
+import { WebSocket, type WebSocketServer } from 'ws';
+import semver from 'semver';
+
+import { TerminalProviderSeverity, TerminalStreamWritable, type ITerminal } from '@rushstack/terminal';
+import { Executable, FileSystem } from '@rushstack/node-core-library';
+
+/**
+ * Allowed Playwright browser names.
+ * @alpha
+ */
+export type BrowserNames = 'chromium' | 'firefox' | 'webkit';
+const validBrowserNames: Set<string> = new Set(['chromium', 'firefox', 'webkit']);
+function isValidBrowserName(browserName: string): browserName is BrowserNames {
+  return validBrowserNames.has(browserName);
+}
+
+/**
+ * Status values reported by {@link PlaywrightTunnel}.
+ * @alpha
+ */
+export type TunnelStatus =
+  | 'waiting-for-connection'
+  | 'browser-server-running'
+  | 'stopped'
+  | 'setting-up-browser-server'
+  | 'error';
+
+interface IHandshake {
+  action: 'handshake';
+  browserName: BrowserNames;
+  launchOptions: LaunchOptions;
+  playwrightVersion: semver.SemVer;
+}
+
+type ITunnelMode = 'poll-connection' | 'wait-for-incoming-connection';
+
+/**
+ * Options for configuring a {@link PlaywrightTunnel} instance.
+ * @alpha
+ */
+export type IPlaywrightTunnelOptions = {
+  terminal: ITerminal;
+  onStatusChange: (status: TunnelStatus) => void;
+  tmpPath: string;
+} & (
+  | {
+      mode: 'poll-connection';
+      wsEndpoint: string;
+    }
+  | {
+      mode: 'wait-for-incoming-connection';
+      listenPort: number;
+    }
+);
+
+interface IBrowserServerProxy {
+  browserServer: BrowserServer;
+  client: WebSocket;
+}
+
+type ISupportedBrowsers = 'chromium' | 'firefox' | 'webkit';
+
+async function sleep(ms: number): Promise<void> {
+  return new Promise((resolve) => setTimeout(resolve, ms));
+}
+
+/**
+ * Hosts a Playwright browser server and forwards traffic over a WebSocket tunnel.
+ * @alpha
+ */
+export class PlaywrightTunnel {
+  private readonly _terminal: ITerminal;
+  private readonly _onStatusChange: (status: TunnelStatus) => void;
+  private readonly _playwrightBrowsersInstalled: Set<string> = new Set();
+  private _status: TunnelStatus = 'stopped';
+  private _initWsPromise?: Promise<WebSocket>;
+  private _keepRunning: boolean = false;
+  private _ws?: WebSocket;
+  private _mode: ITunnelMode;
+  private readonly _wsEndpoint?: string;
+  private readonly _listenPort?: number;
+  private readonly _tmpPath: string;
```

Comment:
Group all the `readonly` properties together and the mutable properties together.